### DISCUSSION TITLE
Table 表格 - 动态对象 ，从后端查询列明，列名没有正常显示

### DISCUSSION THREAD

**Discussion by yc-2503** [dotnetcore/BootstrapBlazor#4753] [Q&A]
public IEnumerable<string> Columns  作为列名，在 OnInitializedAsync 中从后端异步读取列的配置，就没办法正常显示。
如果改成   public IEnumerable<string> Columns = ["A","B","C"]，就会显示A ,B, C 三列，但是也不会显示后台配置的列

`@page "/trackin"
@using PSMS.Core.MyHttpClient
@using PSMS.DirectView.Contracts
@using PSMS.DirectView.MyHttpClient
@using System.Diagnostics.CodeAnalysis
@rendermode InteractiveAuto
<h3>TrackIn</h3>
<div class="col-12">
    <span class="col-6" style="width: 50%;float:left">
        <BootstrapInputGroup>
            <BootstrapInputGroupLabel DisplayText="ID" Width="100" />
            <BootstrapInput @bind-Value="@Id" />
        </BootstrapInputGroup>
    </span>
    <span class="col-6">
        <Button Color="Color.Primary" Text="查询" style="width: 6rem;margin-left:1rem"></Button>
    </span>
</div>

<div class="col-12">
    @if(Columns != null)
    {
        <Table TItem="CustomDynamicData" OnQueryAsync="OnQueryAsync"
               IsStriped="true" IsBordered="true" ShowToolbar="true" ShowColumnList="true" ShowDefaultButtons="false" ShowRefresh="false" ShowSearch="true">
            <TableColumns>
                @foreach (var element in Columns)
                {
                    <TableColumn Field="@element" FieldName="@element" Text="@element">
                        <Template Context="v">
                            <div>@v.Value</div>
                        </Template>
                    </TableColumn>
                }
            </TableColumns>
        </Table>
    }

</div>

@code {
    [Inject] private WIPMPTROQueryService wipmQueryService { get; set; }
    [Inject] private DirectViewService directViewService { get; set; }

    public IEnumerable<string> Columns = new List<string>();

    private string? Id { get; set; } = "SelectAllOpers";

    /// <summary>
    /// OnInitialized
    /// </summary>
    protected override async Task OnInitializedAsync()
    {
        await base.OnInitializedAsync();
        var tmp = await directViewService.ViewDataAsync(new DirectQueryCmd() { Id = Id });
        Columns = tmp.Columns.Select(x => x.ColumName);

    }

    private async Task<QueryData<CustomDynamicData>> OnQueryAsync(QueryPageOptions options)
    {

        var tmp = await directViewService.ViewDataAsync(new DirectQueryCmd() { Id = Id });

        Columns = tmp.Columns.Select(x=>x.ColumName);

        List<CustomDynamicData> items = new List<CustomDynamicData>();
        // 1
        foreach (var item in tmp.Datas)
        {
            Dictionary<string, string> dts = dts = new Dictionary<string, string>();
            IDictionary<string, object> dic = item as IDictionary<string, object>;
            foreach(var it in tmp.Columns )
            {
                dts.Add(it.ColumName, item[it.ColumName]==null?"":item[it.ColumName].ToString());
            }
            items.Add(new CustomDynamicData( dts));
        }

        return new QueryData<CustomDynamicData>() { Items = items };
    }

}
`

**Comment by ArgoZhang** [comment]
@yc-2503 说实话不知道你要表达啥，能否把问题描述清楚，可以参考 issue 模板。你想要什么或者预期结果是什么，实际结果是什么，提供复现示例工程，不要贴点代码，这对解决问题帮助非常小。

**Reply by yc-2503** [reply]
不好意思，没有描述清楚。
我主要是想参考 表格组件的 [动态对象](https://www.blazor.zone/table/dynamic-object)
能够根据配置，去显示列。

我希望是在 OnInitializedAsync 中从后端 读取 Table 的所有列。
但实际结果，OnInitializedAsync 中，在异步方法之后对于 Table 列的修改都没有产生作用。
比如：


```
    public IEnumerable Columns = ["A","B","C"]
    protected override async Task OnInitializedAsync()
    {
        await base.OnInitializedAsync();
        var tmp = await directViewService.ViewDataAsync(new DirectQueryCmd() { Id = Id }); 
        Columns = tmp.Columns.Select(x => x.ColumName); // 将 Columns 改为 ["D","E","F"]
    }
```

页面上显示的 列名依旧是 ["A","B","C"]。

我也怀疑是否是我后端的返回值有问题，所以我也尝试把代码改成


```
    public IEnumerable Columns = ["A","B","C"]
    protected override async Task OnInitializedAsync()
    {
        await base.OnInitializedAsync();
        var tmp = await directViewService.ViewDataAsync(new DirectQueryCmd() { Id = Id }); 
        Columns =  ["D","E","F"]
    }
```

页面上显示的 列名依旧是 ["A","B","C"]。

但是，如果，我将代码改成

```
    public IEnumerable Columns = ["A","B","C"]
    protected override async Task OnInitializedAsync()
    {
        Columns =  ["D","E","F"]
        await base.OnInitializedAsync();
        var tmp = await directViewService.ViewDataAsync(new DirectQueryCmd() { Id = Id }); 
    }
```

页面上显示的 列名就会变成 ["D","E","F"]。

所以我现在的做法是，先将 Columns 定义为 null ，当后端成功查询到Columns 后，再渲染 Table

```
    @if(Columns != null)
    {
        <Table TItem="CustomDynamicData" OnQueryAsync="OnQueryAsync"
               IsStriped="true" IsBordered="true" ShowToolbar="true" ShowColumnList="true" ShowDefaultButtons="false" ShowRefresh="false" ShowSearch="true">
            <TableColumns>
                @foreach (var element in Columns)
                {
                    <TableColumn Field="@element" FieldName="@element" Text="@element">
                        <Template Context="v">
                            <div>@v.Value</div>
                        </Template>
                    </TableColumn>
                }
            </TableColumns>
        </Table>
    }
```

我后续补充复现代码。:)
